r1cs/gadgets/std/src/fields/cmp.rs

```diff
@@ -187,180 +188,351 @@ impl<F: PrimeField> FpGadget<F> {
-        &self,
-        mut cs: CS,
-        other: &FpGadget<F>,
-    ) -> Result<(), SynthesisError> {
-        self.enforce_smaller_or_equal_than_mod_minus_one_div_two(
-            cs.ns(|| "self smaller or equal mod"),
-        )?;
-        other.enforce_smaller_or_equal_than_mod_minus_one_div_two(
-            cs.ns(|| "other smaller or equal mod"),
-        )?;
-        self.enforce_smaller_than_unchecked(cs.ns(|| "enforce smaller unchecked"), other)
+    pub fn enforce_smaller_than_unchecked<CS: ConstraintSystemAbstract<F>>(&self, mut cs: CS, other: &Self) -> Result<(), SynthesisError> {
+        let is_smaller = self.is_smaller_than_unchecked(cs.ns(|| "is smaller unchecked"), other)?;
+        is_smaller.enforce_equal(cs.ns(|| "enforce smaller than"), &Boolean::constant(true))
     }
 
-    /// Helper function to enforce `self < other`. This function assumes `self`
-    /// and `other` are `<= (p-1)/2` and does not generate constraints to
-    /// verify that.
-    fn enforce_smaller_than_unchecked<CS: ConstraintSystemAbstract<F>>(
+    // Variants of cmp functions that assume `self` and `other` are `<= (p-1)/2` and
+    // do not generate constraints to verify that.
+    implement_cmp_functions_variants!(unchecked);
+    // Variants of cmp functions that require `self` and `other` are `<= (p-1)/2` and
+    // impose constraints to verify that.
+    implement_cmp_functions_variants!(restricted);
+
+}
+
+impl<ConstraintF: PrimeField> ComparisonGadget<ConstraintF> for FpGadget<ConstraintF> {
+    /// Output a Boolean that is true iff `self` < `other`. Here `self` and `other`
+    /// can be arbitrary field elements, they are not constrained to be at most (p-1)/2
+    fn is_smaller_than<CS:ConstraintSystemAbstract<ConstraintF>>(
         &self,
         mut cs: CS,
-        other: &FpGadget<F>,
-    ) -> Result<(), SynthesisError> {
-        let is_smaller_than = self.is_smaller_than_unchecked(cs.ns(|| "is smaller"), other)?;
-        //println!("{} Is smaller then {}: {}", self.get_value().unwrap(), other.get_value().unwrap(), is_smaller_than.get_value().unwrap());
-        let lc_one = CS::one();
-        cs.enforce(
-            || "Enforce smaller then",
-            |lc| lc + is_smaller_than.lc(CS::one(), F::one()),
-            |lc| lc + lc_one.clone(),
-            |lc| lc + lc_one,
-        );
-        Ok(())
+        other: &Self,
+    ) -> Result<Boolean, SynthesisError> {
+        let self_bits = self.to_bits_strict(cs.ns(|| "first op to bits"))?;
+        let other_bits = other.to_bits_strict(cs.ns(|| "second op to bits"))?;
+        // extract the least significant MODULUS_BITS-2 bits and convert them to a field element,
+        // which is necessarily lower than (p-1)/2
+        let fp_for_self_lsbs = FpGadget::<ConstraintF>::from_bits(cs.ns(|| "pack second op MSBs"), &self_bits[2..])?;
+        let fp_for_other_lsbs = FpGadget::<ConstraintF>::from_bits(cs.ns(|| "pack second op LSBs"), &other_bits[2..])?;
```

Comment:
Since to_bits_strict checks that the numbers are below the modulus (<= p-1), we can cut only the first bit and already have numbers less than (p-1)/2.

Reply:
This is true if we extract the MODULUS_BITS-1 most significant bits rather than the least significant ones. Nonetheless, it turned out to be possible to implement the function with less constraints also by extracting the MSBs. We should save 4 constraints with this approach. Nice catch!

Reply:
Cool!